File: src/main/kotlin/tads/dipas/server/image/sapl/Sample.java
package tads.dipas.server.image.sapl;

import tads.dipas.server.image.img.Image;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "sample")
public class Sample {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id", nullable = false)
    private Long id;
    @Transient
    private int[][][] image;
    @OneToMany(mappedBy = "sample")
    private List<Seedling> seedlings;
    private String path;
    private String file;
    private double metric;

    public Sample(String path) {
        this.image = Image.imRead(path);
        this.path = path;
        seedlings = new ArrayList<>();
    }

    public Sample() {

    }

    public void setSeedling(Seedling seedling) {
        this.seedlings.add(seedling);
    }

    public Seedling getSeedling(int i) {
        return seedlings.get(i);
    }

    public String getFile() {
        return file;
    }

    public void setFile(String file) {
        this.file = file;
    }

    public List<Seedling> getSeedlings() {
        return seedlings;
    }

    public int[][][] getImage() {
        return image;
    }

    public void setImage(int[][][] imagem) {
        this.image = imagem;
    }

    public String getPath() {
        return path;
    }

    public void setPath(String caminho) {
        this.path = caminho;
    }

    public double getMetric() {
        return metric;
    }

    public void setMetric(double metric) {
        this.metric = metric / 10;
    }

    public String getFilename() {
        String[] split = this.file.split("\\.");
        return split[0];
    }

    public double tallestSeedling() {
        double maior = 0;
        for (Seedling p : this.seedlings) {
            if (p.getTotalLength() > maior)
                maior = p.getTotalLength();
        }
        return maior;
    }
}

File: src/main/kotlin/tads/dipas/server/models/Sample.kt
package tads.dipas.server.models

import lombok.Data
import org.springframework.beans.factory.annotation.Autowired
import tads.dipas.server.services.DescriptorService
import javax.persistence.*

@Data
@javax.persistence.Entity
class Sample (
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id", nullable = false)
    var id: Long? = null,
    @OneToMany
    val entities:List<Entity>? = null,
    @OneToMany
    val inputs:List<Descriptor>? = null,
    @OneToMany
    var images:List<Image> = ArrayList(),
    val software:Int = 0,
)

class SampleRequest(
    private val software:Int,
    private val inputs:List<DescriptorRequest>
){
    fun build(): Sample{
        val list = ArrayList<Descriptor>()
        inputs.forEach{
            list.add(it.build())
        }
        return Sample(software = software, inputs = list)
    }
}

class SampleResponse(
    var id: Long? = null,
    val entities:List<Entity>? = null,
    val inputs:List<Descriptor>? = null,
    var images:List<ImageResponse> = ArrayList(),
    val software:Int = 0,
){
    constructor(sample: Sample) : this(sample.id, sample.entities, sample.inputs, sample.images.map { image -> ImageResponse(image) }, sample.software)
}